File: src/Solutions/Euler571.hs
module Solutions.Euler571 where
import Utils.List(permutations)
import Utils.Numeric(digitsInBase)

-- Execution time: 2208.0214899s

base = 12

toNum :: [Int] -> Int
toNum = foldl (\x y -> x * base + y) 0

-- We are checking down to base 4, since if number n is pandigital in base b^2
-- it is also pandigital in base b.

euler571 :: IO String
euler571 = return $ show $ sum $ take 10 $ filter (\n -> all (all (>0) . digitsInBase n) [base - 1, base - 2..4])
                  $ map toNum $ filter ((>0) . head) $ permutations [0..base - 1]

File: rust/src/solutions/euler571.rs
use crate::utils::numeric::digits::Digits;
use itertools::Itertools;

pub fn euler571() -> String {
    let base: usize = 12;
    (0..base)
        .permutations(base)
        .filter(|perm| perm[0] > 0)
        .map(|perm| perm.into_iter().fold(0, |num, d| num * base + d))
        .filter(|n| {
            (4..base).rev().all(|base| {
                n.digits_count_in_radix(base)
                    .into_iter()
                    .all(|count| count > 0)
            })
        })
        .take(10)
        .sum::<usize>()
        .to_string()
}
